### PR TITLE
[Enhancement] Remove partition directory with retry in shared data clusters

### PR DESCRIPTION
**Pull request by sduzh**
## Why I'm doing:
In shared-data mode, drop partiton just removes the tablet metadata from FE, it does not remove the partition's data. The data deletion relies on staros's shard diff, which will delete the data via `LakeService::delete_tablet` when it finds a unused shard. But `LakeService::delete_tablet` is a very expensive and inefficient way of deleting data, a more efficient way is to find the partition directory and delete the whole directory directly when drop a partition.

## What I'm doing:
Put the dropped partition into the recycle bin, whether it is a force dropped partition or not, the recycle bin will delete the partition's directory before cleaning up the partition, and if the deletion fails, it will keep retrying.

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [x] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java [hunk 1]
```diff
@@ -1108,200 +1078,10 @@ public void readFields(DataInput in) throws IOException {
         }
     }
 
-    private abstract static class RecyclePartitionInfo implements Writable {
```
Comment (sduzh) [inline fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java:1111]: Moved this class to package `com.starrocks.catalog`

File: fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java [hunk 2]
```diff
@@ -1108,200 +1078,10 @@ public void readFields(DataInput in) throws IOException {
...
-            return isInMemory;
-        }
-
-        public void setDbId(long dbId) {
-            this.dbId = dbId;
-        }
-    }
-
-    // only for RangePartition
-    public static class RecyclePartitionInfoV1 extends RecyclePartitionInfo {
```
Comment (sduzh) [inline fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java:1169]: Moved this inner class to package `com.starrocks.catalog`

File: fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java [hunk 3]
```diff
@@ -1108,200 +1078,10 @@ public void readFields(DataInput in) throws IOException {
...
-            tableId = in.readLong();
-            partition = Partition.read(in);
-            range = RangeUtils.readRange(in);
-            dataProperty = DataProperty.read(in);
-            replicationNum = in.readShort();
-            isInMemory = in.readBoolean();
-        }
-    }
-
-    public static class RecyclePartitionInfoV2 extends RecyclePartitionInfo {
```
Comment (sduzh) [inline fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java:1208]: Moved this inner class to package `com.starrocks.catalog`

File: fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java [hunk 4]
```diff
@@ -1108,200 +1078,10 @@ public void readFields(DataInput in) throws IOException {
...
-
-        @Override
-        public void write(DataOutput out) throws IOException {
-            out.writeLong(-1L);
-            String json = GsonUtils.GSON.toJson(this);
-            Text.writeString(out, json);
-        }
-    }
-
-    public static class RecycleRangePartitionInfo extends RecyclePartitionInfoV2 implements GsonPreProcessable,
```
Comment (sduzh) [inline fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java:1236]: Moved this inner class to package `com.starrocks.catalog`

File: fe/fe-core/src/main/java/com/starrocks/persist/gson/GsonUtils.java [hunk 5]
```diff
@@ -280,10 +284,13 @@ public class GsonUtils {
...
-    private static final RuntimeTypeAdapterFactory<CatalogRecycleBin.RecyclePartitionInfoV2>
+    private static final RuntimeTypeAdapterFactory<RecyclePartitionInfoV2>
             RECYCLE_PARTITION_INFO_V_2_ADAPTER_FACTORY
-            = RuntimeTypeAdapterFactory.of(CatalogRecycleBin.RecyclePartitionInfoV2.class, "clazz")
-            .registerSubtype(CatalogRecycleBin.RecycleRangePartitionInfo.class, "RecycleRangePartitionInfo");
+            = RuntimeTypeAdapterFactory.of(RecyclePartitionInfoV2.class, "clazz")
+            .registerSubtype(RecycleRangePartitionInfo.class, "RecycleRangePartitionInfo")
+            .registerSubtype(RecycleLakeRangePartitionInfo.class, "RecycleLakeRangePartitionInfo")
+            .registerSubtype(RecycleListPartitionInfo.class, "RecycleListPartitionInfo")
+            .registerSubtype(RecycleLakeListPartitionInfo.class, "RecycleLakeListPartitionInfo");
```
Comment (gengjun-git) [inline fe/fe-core/src/main/java/com/starrocks/persist/gson/GsonUtils.java:293]: The deserialize of RecyclePartitionInfo in CatalogRecycleBin.read() should use RecyclePartitionInfoV2.class, current is RecycleRangePartitionInfo.class

File: fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java [hunk 6]
```diff
@@ -263,14 +253,7 @@ public synchronized short getPartitionReplicationNum(long partitionId) {
...
         RecyclePartitionInfo partitionInfo = idToPartition.get(partitionId);
-        if (partitionInfo != null) {
-            if (partitionInfo instanceof RecyclePartitionInfoV1) {
-                return ((RecyclePartitionInfoV1) partitionInfo).getRange();
-            } else {
-                return ((RecycleRangePartitionInfo) partitionInfo).getRange();
-            }
-        }
-        return null;
+        return partitionInfo != null ? partitionInfo.getRange() : null;
```
Comment (gengjun-git) [inline fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java:256]: All the getPartition related function needs to filter the unrecoverable partition.

**Review by nshangyiming — DISMISSED** [review]
LGTM

**Review by gengjun-git — APPROVED** [review]

**Review by nshangyiming — APPROVED** [review]

**Review by wyb — APPROVED** [review]

**Pull request merged**